### PR TITLE
Issue #8513: New check RecordTypeParameterNameCheck

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #8513: New check RecordTypeParameterNameCheck

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/e3988f8092d919b5cbe70f5359c4d9e8/raw/a0f6f5860f8025c19868061dee6e0e40960b992f/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/nmancus1/fe2c1055db690af06582ed78d9f52dd5/raw/149357483340a42b902ed1715e005cafb90c3a69/RecordTypeParameterNameCheckBase.xml

Diff Regression patch config: https://gist.githubusercontent.com/nmancus1/ed1d63e8e53dff56fd53642c43a66c76/raw/81408cd27482b5ea8a9e7746d5e7e0eb98c96016/RecordTypeParameterNameCheckPatch.xml

Contribution PR: https://github.com/checkstyle/contribution/pull/499

Link to site: https://nmancus1.github.io/site_RecordTypeParameterName
_updated 8/3 per https://github.com/checkstyle/checkstyle/pull/8586#discussion_r464095811_

~~Reviewers, please note that I included the commit from #8516 PR, since I get an OuterTypeFilenameCheck violation for two input files without it.  I didn't want to add a suppression, etc. which would have to be removed later anyway. I hope that's ok.~~

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

File: .ci/jsoref-spellchecker/whitelist.words [hunk 1]
```diff
@@ -939,6 +939,7 @@ nonvalidating
 noparentfile
 noscript
 nospace
+notfoo
```
Comment (romani) [inline .ci/jsoref-spellchecker/whitelist.words:942]: Can we avoid this ?

File: src/xdocs/config_naming.xml [hunk 2]
```diff
@@ -2437,6 +2437,102 @@ class MyClass {
       </subsection>
     </section>
 
+    <section name="RecordTypeParameterName">
```
Comment (romani) [inline src/xdocs/config_naming.xml:2440]: Please generate web site and share link to it to see rendering.

Comment (nmancus1) [reply src/xdocs/config_naming.xml:2440]: https://nmancus1.github.io/site_ RecordTypeParameterName

File: .ci/jsoref-spellchecker/whitelist.words (hunk 1 above)
Comment (nmancus1) [reply .ci/jsoref-spellchecker/whitelist.words]: Fixed.

File: src/xdocs/config_naming.xml (hunk 2 above)
Comment (romani) [reply src/xdocs/config_naming.xml:2440]: Please investigate and fix indentation on code blocks in html web site, please review other code samples

**Comment by romani** [comment]
I just curious why we need new Check on this ?

How naming model should be different with other types ?

**Comment by pbludov** [comment]
> I just curious why we need new Check on this ?
> How naming model should be different with other types ?

Probably some kind of prefix. That is why we have `MethodTypeParameterName`, `ClassTypeParameterName`, `InterfaceTypeParameterName`.

File: src/xdocs/config_naming.xml (hunk 2 above)
Comment (nmancus1) [reply src/xdocs/config_naming.xml:2440]: Fixed in this PR, and [here](https://github.com/checkstyle/checkstyle/pull/8596), too.

File: pom.xml [hunk 3]
```diff
@@ -1911,6 +1911,7 @@
                 <param>
                   com.puppycrawl.tools.checkstyle.checks.naming.PatternVariableNameCheck*
                 </param>
+                <param>com.puppycrawl.tools.checkstyle.checks.naming.RecordTypeParameterNameCheck*</param>
```
Comment (rnveach) [inline pom.xml:1914]: Why is this naming check added to the pitest-misc group?

File: pom.xml [hunk 4]
```diff
@@ -1911,6 +1911,7 @@
                 <param>
                   com.puppycrawl.tools.checkstyle.checks.naming.PatternVariableNameCheck*
```
Comment (rnveach) [inline pom.xml:1912]: I don't think this check should be here too.

File: pom.xml (hunk 3 above)
Comment (nmancus1) [reply pom.xml:1914]: Fixed.

File: pom.xml (hunk 4 above)
Comment (nmancus1) [reply pom.xml:1912]: https://github.com/checkstyle/checkstyle/pull/8649

**Pull request merged**